clesperanto/core/plugin_function.py

```diff
@@ -0,0 +1,105 @@
+import inspect
+from typing import Any, Callable, Dict, Optional, Sequence, Set, Type, Union
+from functools import wraps
+from toolz import curry
+
+from gputools import OCLArray
+
+from .create import create_like
+from .push import push
+
+@curry
+def plugin_function(
+    function: Callable,
+    output_creator: Callable = create_like
+) -> Callable:
+    """Function decorator which ensures correct types and values of parameters
+        Given input parameters are either of type OCLArray (which the GPU understands) or are converted to this type
+        (see push function). If output parameters of type OCLArray are not set, an empty image is created and handed
+        over.
+
+    :param function: the function to be called on the GPU
+    :param output_creator: a function which can create an output OCLArray given an input OCLArray, optional
+        per default, we create output images of the same shape as input images
+    :return: return value of the executed function
+    """
+
+
+    @wraps(function)
+    def worker_function(*args, **kwargs):
+        # determine argument spec and default values, values are given as args
+        argument_specification = inspect.getfullargspec(function)
+        defaults_values = argument_specification.defaults
+
+        target_arguments = [None] * len(argument_specification.args)
+
+        arg_counter = 0
+        default_counter = 0
+        any_input = None
+        for argument in argument_specification.args:
+            #print("---\nparsing argument " + argument)
+            if (len(args) > arg_counter):
+                value = args[arg_counter]
+                #print("value")
+                #print(value)
+            else:
+                value = None
+
+            # was the argument annotated?
+            type_annotation = argument_specification.annotations.get(argument);
+            if (type_annotation is not None):
```

Comment:
@haesleinhuepf I guess we should skip the type annotation-based dispatch, since the annotation will be *either* NumPy or OCL.

Another point is that we could make `push` itself magical and be a no-op for OCLArrays.

So with all the suggestions I've made so far I think you're getting rid of almost all the if/elses? :joy:

Reply:
You might right. [Push(x) a.ka. convert(x, OCLArray)](https://github.com/clij/clij-core/blob/888ab4c5484477d3192bdb37425bdf94a4699ce4/src/main/java/net/haesleinhuepf/clij/CLIJ.java#L449) in clij is also magic. Let's see how many IFs I can spare...